plugins/MauticFullContactBundle/EventListener/LeadSubscriber.php

```diff
@@ -0,0 +1,160 @@
+use MauticPlugin\MauticFullContactBundle\Services\FullContact_Person;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+class LeadSubscriber extends CommonSubscriber
+{
+    /**
+     * @var IntegrationHelper
+     */
+    protected $integrationHelper;
+
+    /**
+     * @var UserHelper
+     */
+    protected $userHelper;
+
+    /**
+     * @var LeadModel
+     */
+    protected $leadModel;
+
+    /**
+     * @var CompanyModel
+     */
+    protected $companyModel;
+
+    /**
+     * LeadSubscriber constructor.
+     *
+     * @param IntegrationHelper $integrationHelper
+     * @param UserHelper        $userHelper
+     * @param LeadModel         $leadModel
+     * @param CompanyModel      $companyModel
+     */
+    public function __construct(IntegrationHelper $integrationHelper, UserHelper $userHelper, LeadModel $leadModel, CompanyModel $companyModel)
+    {
+        parent::__construct();
+
+        $this->integrationHelper = $integrationHelper;
+        $this->userHelper        = $userHelper;
+        $this->leadModel         = $leadModel;
+        $this->companyModel      = $companyModel;
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return [
+            LeadEvents::LEAD_POST_SAVE    => ['leadPostSave', 0],
+            LeadEvents::COMPANY_POST_SAVE => ['companyPostSave', 0],
+        ];
+    }
+
+    public function leadPostSave(LeadEvent $event)
+    {
+        /** @var FullContactIntegration $myIntegration */
+        $myIntegration = $this->integrationHelper->getIntegrationObject('FullContact');
+
+        if (!$myIntegration->getIntegrationSettings()->getIsPublished()) {
+            return;
+        }
+
+        $lead = $event->getLead();
+
+        // get api_key from plugin settings
+        $keys = $myIntegration->getDecryptedApiKeys();
+
+        if ($myIntegration->shouldAutoUpdate()) {
+            $fullcontact = new FullContact_Person($keys['apikey']);
+            try {
+                /** @var User $user */
+                $user      = $this->userHelper->getUser();
+                $webhookId = 'fullcontact_notify#'.$lead->getId().'#'.$user->getId();
+                $cache     = $lead->getSocialCache() ?: [];
+                $cacheId   = sprintf('%s%s', $webhookId, date(DATE_ATOM));
+                if (!array_key_exists($cacheId, $cache)) {
+                    $fullcontact->setWebhookUrl(
+                        $this->router->generate(
+                            'mautic_plugin_fullcontact_index',
+                            [],
+                            UrlGeneratorInterface::ABSOLUTE_URL
+                        ),
+                        $webhookId
+                    );
+                    $res             = $fullcontact->lookupByEmailMD5(md5($lead->getEmail()));
+                    $cache[$cacheId] = serialize($res);
+                    $lead->setSocialCache($cache);
+                    $this->leadModel->saveEntity($lead);
```

Comment:
This is causing a loop because saveEntity will dispatch LeadEvents::LEAD_POST_SAVE again so it gets called over and over and over. Rather than `$this->leadModel->saveEntity($lead);` try using `$this->leadModel->getRepository()->saveEntity($lead);` so that it doesn't re-trigger the event.